contracts/container/contract.go

```diff
@@ -491,6 +493,100 @@ func List(owner []byte) [][]byte {
 	return list
 }
 
+// AddNextEpochNodes accumulates passed nodes as container members for the next
+// epoch to be committed using [CommitContainerListUpdate]. Results of the call
+// operation can be received via [Nodes]. This method must be called only when
+// a container list is changed, otherwise nothing should be done.
+// Call must be signed by the Alphabet nodes.
+func AddNextEpochNodes(cID interop.Hash256, publicKeys []interop.PublicKey) {
+	if len(cID) != interop.Hash256Len {
+		panic(cst.ErrorInvalidContainerID + ": length: " + std.Itoa10(len(cID)))
+	}
+
+	ctx := storage.GetContext()
+	multiaddr := common.AlphabetAddress()
+	common.CheckAlphabetWitness(multiaddr)
+
+	counter := 0
+	counterKey := append([]byte{nextEpochNodesPrefix, 'i'}, cID...)
+	if counterRaw := storage.Get(ctx, counterKey); counterRaw != nil {
+		counter = counterRaw.(int)
+	}
+
+	commonPrefix := append([]byte{nextEpochNodesPrefix}, cID...)
+	for _, publicKey := range publicKeys {
+		if len(publicKey) != interop.PublicKeyCompressedLen {
+			panic(cst.ErrorInvalidPublicKey + ": length: " + std.Itoa10(len(publicKey)))
+		}
+
+		counter++
+
+		storageKey := append(commonPrefix, std.Serialize(counter)...)
+		storage.Put(ctx, storageKey, publicKey)
+	}
+
+	storage.Put(ctx, counterKey, counter)
+}
+
+// CommitContainerListUpdate commits container list changes made by
+// [AddNextEpochNodes] calls in advance. If no [AddNextEpochNodes] have been
+// made, it clears container list. Makes "ContainerUpdate" notification with
+// container ID after successful list change.
+// Call must be signed by the Alphabet nodes.
+func CommitContainerListUpdate(cID interop.Hash256) {
+	if len(cID) != interop.Hash256Len {
+		panic(cst.ErrorInvalidContainerID + std.Itoa10(len(cID)))
+	}
+
+	ctx := storage.GetContext()
+	multiaddr := common.AlphabetAddress()
+	common.CheckAlphabetWitness(multiaddr)
+
+	oldCounterKey := append([]byte{nodesPrefix, 'i'}, cID...)
+	oldNodesPrefix := append([]byte{nodesPrefix}, cID...)
+	newCounterKey := append([]byte{nextEpochNodesPrefix, 'i'}, cID...)
+	newNodesPrefix := append([]byte{nextEpochNodesPrefix}, cID...)
+
+	oldNodes := storage.Find(ctx, oldNodesPrefix, storage.KeysOnly)
+	for iterator.Next(oldNodes) {
+		oldNode := iterator.Value(oldNodes).(string)
+		storage.Delete(ctx, oldNode)
+	}
+	storage.Delete(ctx, oldCounterKey)
+
+	newNodes := storage.Find(ctx, newNodesPrefix, storage.None)
+	for iterator.Next(newNodes) {
+		newNode := iterator.Value(newNodes).(struct {
+			key []byte
+			val []byte
+		})
+
+		storage.Delete(ctx, newNode.key)
+
+		newKey := append([]byte{nodesPrefix}, newNode.key[1:]...)
+		storage.Put(ctx, newKey, newNode.val)
+	}
+	if newCounter := storage.Get(ctx, newCounterKey); newCounter != nil {
+		storage.Put(ctx, oldCounterKey, storage.Get(ctx, newCounterKey))
+	} else {
+		storage.Delete(ctx, newCounterKey)
```

Comment:
dropped eventually

> and shouldnt we always reset it?

but why not? meaningless counter, yes it may be overwritten with the next container update but before it happenes it is just a garbage